### PR TITLE
SNOW-895537: Send query context with request

### PR DESCRIPTION
**Pull request by sfc-gh-pfus**
### Description
Query contexts are now sent to snowflake backend.

### Checklist
- [x] Code compiles correctly
- [x] Run ``make fmt`` to fix inconsistent formats
- [x] Run ``make lint`` to get lint errors and fix all of them
- [x] Created tests which fail without the change (if possible)
- [x] All tests passing
- [x] Extended the README / documentation, if necessary

### CODE REVIEW

File: connection.go [hunk 1]
```diff
@@ -85,13 +86,18 @@ func (sc *snowflakeConn) exec(
 	var err error
 	counter := atomic.AddUint64(&sc.SequenceCounter, 1) // query sequence counter
 
+	queryContext, err := buildQueryContext(sc.queryContextCache)
+	if err != nil {
+		logger.Warnf("error while building query context: %v", err)
```
Comment (sfc-gh-dheyman-1) [inline connection.go:91]: I think this should be error.

File: connection.go [hunk 2]
```diff
@@ -160,6 +166,45 @@ func (sc *snowflakeConn) exec(
 	return data, err
 }
 
+func buildQueryContext(qcc *queryContextCache) (requestQueryContext, error) {
```
Comment (sfc-gh-dheyman-1) [inline connection.go:169]: don't we want to extract these two methods to another file? connection.go is getting rather large.

File: connection.go [hunk 3]
```diff
@@ -160,6 +166,45 @@ func (sc *snowflakeConn) exec(
...
+			Timestamp: qce.Timestamp,
+			Context: contextData{
+				Base64Data: contextBase64,
+			},
+		})
+	}
+	return rqc, nil
+}
+
+func marshallContext(context any) (string, error) {
```
Comment (sfc-gh-dheyman-1) [inline connection.go:192]: `marshalContext` EDIT: actually a matter of US spelling, so no need to change.

**Review by sfc-gh-dheyman-1 — APPROVED** [review]

**Review by sfc-gh-dstempniak — APPROVED** [review]

File: connection.go (hunk 1 above)
Comment (sfc-gh-pfus) [reply connection.go]: Right, done.

File: connection.go (hunk 2 above)
Comment (sfc-gh-pfus) [reply connection.go:169]: This one is pretty internal. But I agree, that we should reorganise files a bit, but in different commit.

File: htap_test.go [hunk 4]
```diff
@@ -416,3 +419,79 @@ func htapTestSnowflakeConn() *snowflakeConn {
...
+		})
+		t.Run("Run tests on first database again", func(t *testing.T) {
+			sct.mustExec(fmt.Sprintf("USE DATABASE %v", testDb1), nil)
+
+			sct.mustExec("INSERT INTO test_hybrid_table VALUES (4, 'd')", nil)
+
+			rows := sct.mustQuery("SELECT * FROM test_hybrid_table", nil)
+			defer rows.Close()
+			if len(sct.sc.queryContextCache.entries) != 3 {
+				t.Errorf("expected two entries in query context cache, got: %v", sct.sc.queryContextCache.entries)
```
[the remark this answers is not in the document]
Comment (sfc-gh-pfus) [reply htap_test.go]: Really appreciate, that someone reads so carefully and finds my easter eggs!

File: connection.go [hunk 5]
```diff
@@ -85,13 +85,18 @@ func (sc *snowflakeConn) exec(
 	var err error
 	counter := atomic.AddUint64(&sc.SequenceCounter, 1) // query sequence counter
 
+	queryContext, err := buildQueryContext(sc.queryContextCache)
+	if err != nil {
+		logger.Errorf("error while building query context: %v", err)
```
Comment (sfc-gh-igarish) [inline connection.go:90]: Do we continue even if error occurred?

Comment (sfc-gh-pfus) [reply connection.go:90]: Yes. This is confirmed with Lianke, that we should proceed and just not send query context.

**Pull request merged**